metals/src/main/scala/scala/meta/internal/metals/StacktraceAnalyzer.scala

```diff
@@ -0,0 +1,227 @@
+package scala.meta.internal.metals
+
+import java.io.FileWriter
+import java.net.URLEncoder
+import java.util.Collections.singletonList
+
+import scala.io.Source
+import scala.util.Try
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+
+import com.google.gson.JsonPrimitive
+import org.eclipse.lsp4j.Location
+import org.eclipse.{lsp4j => l}
+
+class StacktraceAnalyzer(
+    workspace: AbsolutePath,
+    definitionProvider: DefinitionProvider,
+    gotoicon: String,
+    commandsInHtmlSupported: Boolean
+) {
+
+  def analyzeCommand(
+      commandParams: l.ExecuteCommandParams
+  ): Option[l.ExecuteCommandParams] = {
+    parseJsonParams(commandParams)
+      .flatMap(analyzeStackTrace)
+  }
+
+  def matches(path: AbsolutePath): Boolean =
+    path == workspace.resolve(Directories.stacktrace)
+
+  def highlight(
+      path: AbsolutePath,
+      lineNumber: Int
+  ): java.util.List[l.DocumentHighlight] = {
+    val lineOpt = readStacktraceFileLine(path, lineNumber)
+    lineOpt match {
+      case Some(line) =>
+        val from = new l.Position(lineNumber, line.indexOf("at ") + 3)
+        val to = new l.Position(lineNumber, line.indexOf("("))
+        val range = new l.Range(from, to)
+        singletonList(
+          new l.DocumentHighlight(range, l.DocumentHighlightKind.Read)
+        )
+      case None =>
+        java.util.Collections.emptyList()
+    }
+  }
+
+  def stacktraceLenses(path: AbsolutePath): Seq[l.CodeLens] = {
+    stacktraceLenses(readStacktraceFile(path))
+  }
+
+  private def readStacktraceFileLine(
+      path: AbsolutePath,
+      lineNumber: Int
+  ): Option[String] =
+    readStacktraceFile(path).drop(lineNumber).headOption
+
+  private def readStacktraceFile(path: AbsolutePath): List[String] = {
+    Source.fromFile(path.toString).getLines().toList
+  }
+
+  def definition(
+      path: AbsolutePath,
+      position: l.Position
+  ): java.util.List[Location] = {
+    val lineOpt = readStacktraceFileLine(path, position.getLine)
+    lineOpt match {
+      case Some(line) =>
+        getSymbolLocationFromLine(line).toList.asJava
+      case None =>
+        java.util.Collections.emptyList()
+    }
+  }
+
+  private def adjustPosition(lineNumber: Int, pos: l.Position): Unit = {
+    pos.setLine(lineNumber)
+    pos.setCharacter(0)
+  }
+
+  private def tryGetLineNumberFromStacktrace(line: String): Try[Int] = {
+    Try(
+      // number is 1-based, but editors are 0-based
```

Comment:
LSP is the only one that uses 0 based ones. Both editor itself and stacktrace are 1-based. Maybe we could make the comment more precise?

Reply:
updated, not resolving it yet, let me know if it makes more sense

Reply:
`// number is 1-based, but LSP uses 0-based line numbers` ?

Reply:
haha sorry I wrote a comment and then pushed my change, you must have noticed unchanged comment, WDYT about a new one? :)

Reply:
it is mainly the same as you proposed, will mark it as resolved